## 79% of end-end tests have passed
```
Total executed tests: 14
Failed tests: 3
Expected to fail tests: 0
Passed tests: 11
```
```
IDs of failed tests: 741840,741839,741841 
```
<h3>Failed tests (3)</h3><details><summary>Click to expand</summary><br/><li><a href="https://ci.status.im/job/status-mobile/job/e2e/job/status-app-prs-rerun/parambuild/?BRANCH_NAME=chore/next-stage-pin&PR_ID=22216&APK_NAME=22216.apk&TR_CASE_IDS=741840,741839,741841">Rerun failed tests</a></li><br/><h4>Class TestWalletCollectibles:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3804008">test_wallet_send_collectible</a>, id: 741840 </b></td></tr><tr><td><p><blockquote><div>Device 1: Tap on found: CollectibleItemElement </div><div>Device 1: Find `Button` by `accessibility id`: `amount-input-inc-button` </div></blockquote></p>

```
critical/wallet/test_collectibles.py:105: in test_wallet_send_collectible
    self.wallet_view.amount_input_increase_button.click()
../views/base_element.py:89: in click
    element = self.find_element()
../views/base_element.py:78: in find_element
    raise NoSuchElementException(
 Device 1: Button by accessibility id: `amount-input-inc-button` is not found on the screen; For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors#no-such-element-exception
```

<br/><br/></td></tr><tr><td><b>2. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3804007">test_wallet_collectibles_balance</a>, id: 741839 </b></td></tr><tr><td><p><blockquote><div>Device 1: `Text` is `Double Spike` </div><div>Device 1: Click system back button </div></blockquote></p>

```
critical/wallet/test_collectibles.py:94: in test_wallet_collectibles_balance
    self.errors.verify_no_errors()
base_test_case.py:179: in verify_no_errors
    pytest.fail('\n '.join([self.errors.pop(0) for _ in range(len(self.errors))]))
 Device 1: BVL image doesn't match expected template
E    Device 1: BVL quantity 1 doesn't match expected 2
E    Device 1: BVL expanded image doesn't match expected template
```

<br/><br/></td></tr><tr><td><b>3. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3804009">test_wallet_collectible_send_from_expanded_info_view</a>, id: 741841 </b></td></tr><tr><td><p><blockquote><div>Device 1: Find `Button` by `accessibility id`: `collectibles-tab` </div><div>Device 1: Tap on found: Button </div></blockquote></p>

```
critical/wallet/test_collectibles.py:150: in test_wallet_collectible_send_from_expanded_info_view
    self.wallet_view.get_collectible_element('Glitch Punks').wait_for_element().click()
../views/base_element.py:120: in wait_for_element
    raise TimeoutException(
 Device `1`: `CollectibleItemElement` by` xpath`: `//*[@content-desc='collectible-list-item']//*[contains(@text,'Glitch Punks')]/../..` is not found on the screen after wait_for_element
```

<br/><br/></td></tr></tbody></table></details><h3>Passed tests (11)</h3><details><summary>Click to expand</summary><br/><br/><h4>Class TestWalletMultipleDevice:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3804001">test_wallet_send_asset_from_drawer</a>, id: 727230 </b></td></tr><tr><td></td></tr><tr><td><b>2. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3804000">test_wallet_send_eth</a>, id: 727229 </b></td></tr><tr><td></td></tr></tbody></table><h4>Class TestOneToOneChatMultipleSharedDevicesNewUi:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3803997">test_1_1_chat_non_latin_messages_stack_update_profile_photo</a>, id: 702745 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://appautomation.lambdatest.com/test?testID=APPTEST3ZuiZKaJBozNaJL">Steps, video, logs</a></li></ul></p><p>Device 2:<ul><li><a href="https://appautomation.lambdatest.com/test?testID=APPTES6WMUBVUbsjEkGoZd">Steps, video, logs</a></li></ul></p></ul></p></td></tr></tbody></table><h4>Class TestCommunityOneDeviceMerged:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3803996">test_community_copy_and_paste_message_in_chat_input</a>, id: 702742 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://appautomation.lambdatest.com/test?testID=APPTES7m6z7rIhZGMWix1s">Steps, video, logs</a></li></ul></p></ul></p></td></tr><tr><td><b>2. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3803999">test_restore_multiaccount_with_waku_backup_remove_profile_switch</a>, id: 703133 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://appautomation.lambdatest.com/test?testID=APPTES7m6z7rIhZGMWix1s">Steps, video, logs</a></li></ul></p></ul></p></td></tr></tbody></table><h4>Class TestCommunityMultipleDeviceMerged:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3803998">test_community_message_edit</a>, id: 702843 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://appautomation.lambdatest.com/test?testID=APPTES94dmdkZfBp5AfHx9">Steps, video, logs</a></li></ul></p><p>Device 2:<ul><li><a href="https://appautomation.lambdatest.com/test?testID=APPTESQkHIqhOXJUHcnJG0">Steps, video, logs</a></li></ul></p></ul></p></td></tr></tbody></table><h4>Class TestWalletOneDevice:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3804005">test_wallet_swap_flow_mainnet</a>, id: 741555 </b></td></tr><tr><td></td></tr><tr><td><b>2. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3804002">test_wallet_add_remove_regular_account</a>, id: 727231 </b></td></tr><tr><td></td></tr><tr><td><b>3. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3804003">test_wallet_balance_mainnet</a>, id: 740490 </b></td></tr><tr><td></td></tr><tr><td><b>4. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3804006">test_wallet_bridge_flow_mainnet</a>, id: 741612 </b></td></tr><tr><td></td></tr><tr><td><b>5. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3804004">test_wallet_send_flow_mainnet</a>, id: 741554 </b></td></tr><tr><td></td></tr></tbody></table></details>

--- a/src/status_im/contexts/keycard/change_pin/events.cljs
+++ b/src/status_im/contexts/keycard/change_pin/events.cljs
@@ -36,7 +36,7 @@
    {:db (assoc-in db [:keycard :change-pin :new-pin] (security/mask-data pin))}))
 
 (defn- change-pin-and-continue
-  [current-pin new-pin]
+  [change-pin-data]
   (rf/dispatch [:keycard/change-pin
                 {:on-success  (fn []
                                 (rf/dispatch [:navigate-back])
@@ -46,15 +46,18 @@
                                 (rf/dispatch [:navigate-back])
                                 (rf/dispatch [:keycard/disconnect])
                                 (rf/dispatch [:open-modal :screen/keycard.pin-change-failed]))
-                 :current-pin current-pin
-                 :new-pin     new-pin}]))
+                 :current-pin (security/safe-unmask-data (:current-pin change-pin-data))
+                 :new-pin     (security/safe-unmask-data (:new-pin change-pin-data))}]))
 
 (defn- verify-pin-and-continue
-  [current-pin new-pin]
+  [change-pin-data]
   (rf/dispatch
    [:keycard/verify-pin
-    {:pin        current-pin
-     :on-success #(change-pin-and-continue current-pin new-pin)
+    {:pin        (security/safe-unmask-data (:current-pin change-pin-data))
+     :on-success (fn []
+                   (rf/dispatch [:keycard/connect.next-stage
+                                 {:key-uid    (:key-uid change-pin-data)
+                                  :on-success #(change-pin-and-continue change-pin-data)}]))
      :on-failure (fn [error]
                    (when-not (utils/tag-lost? (:error error))
                      (rf/dispatch [:keycard/disconnect])
@@ -63,11 +66,10 @@
 
 (rf/reg-event-fx :keycard/change-pin.verify-current-pin-and-continue
  (fn [{:keys [db]}]
-   (let [{:keys [current-pin new-pin]} (get-in db [:keycard :change-pin])
-         unmasked-current-pin          (security/safe-unmask-data current-pin)
-         unmasked-new-pin              (security/safe-unmask-data new-pin)]
+   (let [key-uid         (get-in db [:profile/profile :key-uid])
+         change-pin-data (assoc (get-in db [:keycard :change-pin]) :key-uid key-uid)]
      {:fx [[:dispatch
             [:keycard/connect
              {:theme      :dark
-              :key-uid    (get-in db [:profile/profile :key-uid])
-              :on-success #(verify-pin-and-continue unmasked-current-pin unmasked-new-pin)}]]]})))
+              :key-uid    key-uid
+              :on-success #(verify-pin-and-continue change-pin-data)}]]]})))

